pkg/virt-controller/services/template.go

```diff
@@ -558,6 +560,33 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, i
 
 	setNodeAffinityForPod(vmi, &pod)
 
+	tscFrequency := nsr.TSCFrequency()
+	// This just takes 250 PPM of the frequency. The /1000 and *1000 is just to round down to the nearest kHz
+	tolerance := int64(math.Floor(float64(tscFrequency/1000)*(float64(topology.TSCTolerancePPM)/1000000)) * 1000)
+	if tscFrequency != 0 {
+		if pod.Spec.Affinity == nil {
+			pod.Spec.Affinity = &k8sv1.Affinity{}
+		}
+		pod.Spec.Affinity.NodeAffinity = &k8sv1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: &k8sv1.NodeSelector{
+				NodeSelectorTerms: []k8sv1.NodeSelectorTerm{{
+					MatchExpressions: []k8sv1.NodeSelectorRequirement{
+						{
+							Key:      topology.TSCFrequencyLabel,
+							Operator: k8sv1.NodeSelectorOpGt,
+							Values:   []string{strconv.FormatInt(tscFrequency-tolerance, 10)},
+						},
+						{
+							Key:      topology.TSCFrequencyLabel,
+							Operator: k8sv1.NodeSelectorOpLt,
+							Values:   []string{strconv.FormatInt(tscFrequency+tolerance, 10)},
+						},
+					},
+				}},
+			},
+		}
+	}
+
```

Comment:
As we're messing with affinities above (`setNodeAffinityForPod()`), we should avoid overriding affinities and instead append the new ones.

```suggestion
	if tscFrequency != 0 {
		if pod.Spec.Affinity == nil {
			pod.Spec.Affinity = &k8sv1.Affinity{}
		}
		if pod.Spec.Affinity.NodeAffinity == nil {
			pod.Spec.Affinity.NodeAffinity = &k8sv1.NodeAffinity{}
		}
		if pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
			pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution = &k8sv1.NodeSelector{}
		}
		
		selector := pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution

		if selector.NodeSelectorTerms == nil {
			selector.NodeSelectorTerms = []k8sv1.NodeSelectorTerm{}
		}
		if selector.NodeSelectorTerms.MatchExpressions == nil {
			selector.NodeSelectorTerms.MatchExpressions = []k8sv1.NodeSelectorRequirement{}
		}

		MatchExpressions := selector.NodeSelectorTerms.MatchExpressions

		MatchExpressions = append(MatchExpressions, k8sv1.NodeSelectorRequirement{
			{
				Key:      topology.TSCFrequencyLabel,
				Operator: k8sv1.NodeSelectorOpGt,
				Values:   []string{strconv.FormatInt(tscFrequency-tolerance, 10)},
			},
			{
				Key:      topology.TSCFrequencyLabel,
				Operator: k8sv1.NodeSelectorOpLt,
				Values:   []string{strconv.FormatInt(tscFrequency+tolerance, 10)},
			},
		})
	}
```

A bit ugly, but necessary.
WDYT?

Reply:
I would suggest to not use lt and gt. If we go with that, we will not target nodes which can scale their TSC frequency. Instead as suggested in https://github.com/kubevirt/kubevirt/pull/9615#discussion_r1170133542 I would "just" slightly change the tsc frequency calculation and silghtly change which nodes get the resulting label added.

This way we can cover nodes which can't scale TSC as well as those which can scale.